src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/RemoteTypes/RemotePackageSourceProvider.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            return NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await _initializer.InitializeAsync(CancellationToken.None);
+                return await _sourceRepositoryService.IsPackageSourceEnabledAsync(name, CancellationToken.None);
+            });
+        }
+
+        public IEnumerable<PackageSource> LoadPackageSources()
+        {
+            return NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await _initializer.InitializeAsync(CancellationToken.None);
+                return await _sourceRepositoryService.LoadPackageSourcesAsync(CancellationToken.None);
+            });
+        }
+
+        public void RemovePackageSource(string name)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await _initializer.InitializeAsync(CancellationToken.None);
+                await _sourceRepositoryService.RemovePackageSourceAsync(name, CancellationToken.None);
+            });
+        }
+
+        public void SaveActivePackageSource(PackageSource source)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await _initializer.InitializeAsync(CancellationToken.None);
+                await _sourceRepositoryService.SaveActivePackageSourceAsync(source, CancellationToken.None);
+            });
+        }
+
+        public void SavePackageSources(IEnumerable<PackageSource> sources)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await _initializer.InitializeAsync(CancellationToken.None);
+                await _sourceRepositoryService.SavePackageSourcesAsync(sources, CancellationToken.None);
+            });
+        }
+
+        public void UpdatePackageSource(PackageSource source, bool updateCredentials, bool updateEnabled)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await _initializer.InitializeAsync(CancellationToken.None);
+                await _sourceRepositoryService.UpdatePackageSourceAsync(source, updateCredentials, updateEnabled, CancellationToken.None);
+            });
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void OnPackageSourcesChanged(object sender, EventArgs e)
+        {
+            PackageSourcesChanged?.Invoke(this, e);
+        }
+        private async Task InitializeAsync()
+        {
+            if (_initialized)
+            {
+                return;
+            }
+
+            var remoteBroker = await BrokeredServicesUtilities.GetRemoteServiceBroker();
+            _sourceRepositoryService = await remoteBroker.GetProxyAsync<INuGetSourceRepositoryService>(NuGetBrokeredServices.SourceRepositoryProviderService);
```

Comment:
Any time you acquire and store a proxy in a field, you should consider hooking the `AvailabilityChanged` event and refreshing your proxy. For example if the user switches from a local folder to joining a Live Share session, the proxy you got before may no longer apply and you'd need to refresh it.